types/opt/value.go

```diff
@@ -0,0 +1,166 @@
+type equatable[T any] interface {
+	// Equal reports whether the receiver and the other values are equal.
+	Equal(other T) bool
+}
+
+// Empty returns an empty optional Value.
+func Empty[T any]() Value[T] {
+	return Value[T]{set: false}
+}
+
+// Of returns an optional Value containing the specified value.
+// It treats nil pointers, unsafe pointers, and interfaces as unset values,
+// but nil slices and maps as empty slices and maps.
+func Of[T any](v T) Value[T] {
+	switch rv := reflect.ValueOf(v); rv.Kind() {
+	case reflect.Invalid:
+		// A nil interface (e.g. `any(nil)`) is equivalent to an unset value.
+		// It should be marshalled as null, or be omitted from the output.
+		return Empty[T]()
+	case reflect.Pointer, reflect.UnsafePointer, reflect.Interface:
+		if rv.IsNil() {
+			// A nil pointer, unsafe pointer, or interface is equivalent to an unset value.
+			// It should be marshalled as null, or be omitted from the output.
+			return Empty[T]()
+		}
+	case reflect.Map, reflect.Slice:
+		if rv.IsNil() {
+			// A nil map or slice is equivalent to an empty map or slice,
+			// and should be marshalled as `{}` or `[]`, accordingly.
+			return Value[T]{value: v, set: true}
+		}
+	}
+	return Value[T]{value: v, set: true}
+}
+
+// String implements Stringer.
+func (o *Value[T]) String() string {
+	if !o.set {
+		return fmt.Sprintf("(empty[%T])", o.value)
+	}
+	return fmt.Sprint(o.value)
+}
+
+// Set assigns the specified value to the optional value o.
+func (o *Value[T]) Set(v T) {
+	*o = Of(v)
+}
+
+// Clear resets o to an empty state.
+func (o *Value[T]) Clear() {
+	*o = Empty[T]()
+}
+
+// Present reports whether o has a value set.
+func (o *Value[T]) Present() bool {
+	return o.set
+}
+
+// Get returns the value of o, panicking if it hasn't been set.
+func (o Value[T]) Get() T {
+	if !o.set {
+		panic(fmt.Sprintf("%v is not set", reflect.TypeFor[T]()))
+	}
+	return o.value
+}
+
+// Get returns the value and a flag indicating whether the value is set.
+func (o Value[T]) GetOk() (v T, ok bool) {
+	return o.value, o.set
+}
+
+// Equal reports whether o is equal to v.
+// Two optional values are equal if both are empty,
+// or if both are set and the underlying values are equal.
+// If the template type T implements an Equal(T) bool method, it will be used
+// instead of the == operator for value comparison.
+// If T is not comparable, it returns false.
+func (o Value[T]) Equal(v Value[T]) bool {
+	if o.set != v.set {
+		return false
+	}
+	if !o.set {
+		return true
+	}
+	ov := any(o.value)
+	if eq, ok := ov.(equatable[T]); ok {
+		return eq.Equal(v.value)
+	}
+	if reflect.TypeFor[T]().Comparable() {
+		return ov == any(v.value)
+	}
+	return false
+}
+
+// EqualValue reports whether o is equal to v.
+// If the template type T implements an Equal(T) bool method, it will be used
+// instead of the == operator for value comparison.
+// If T is not comparable, it returns false.
+func (o Value[T]) EqualValue(v T) bool {
```

Comment:
Simpler as:
```go
return o.Equal(Of(v))
```
and ensures that the behavior of this and `Equal` are in sync.

Reply:
Actually, do we even need `EqualValue` given how simple this is underneath?

Reply:
Removed.